py/vtdb/vtgatev2.py

```diff
@@ -298,60 +298,91 @@ def _execute_entity_ids(
+        self._update_session(response)
+        if response.reply.get('Error'):
+          raise gorpc.AppError(response.reply['Error'], exec_method)
+        for reply in response.reply['List']:
+          fields = []
+          conversions = []
+          results = []
+          rowcount = 0
+          for field in reply['Fields']:
+            fields.append((field['Name'], field['Type']))
+            conversions.append(field_types.conversions.get(field['Type']))
+          for row in reply['Rows']:
+            results.append(tuple(_make_row(row, conversions)))
+          rowcount = reply['RowsAffected']
+          lastrowid = reply['InsertId']
+          rowsets.append((results, rowcount, lastrowid, fields))
+      except gorpc.GoRpcError as e:
+        self.logger_object.log_private_data(bind_variables_list)
+        raise self._convert_exception(
+            e, sql_list, exec_method,
+            keyspace='', tablet_type=tablet_type)
+      except Exception:
+        logging.exception('gorpc low-level error')
+        raise
+      return rowsets
 
+    # Return merged keyspace_ids and shards rowsets.
     rowsets = []
-
-    try:
-      req = {
-          'Queries': query_list,
-          'TabletType': tablet_type,
-          'AsTransaction': as_transaction,
-      }
-      self._add_caller_id(req, effective_caller_id)
-      self._add_session(req)
-      response = self._get_client().call('VTGate.ExecuteBatchKeyspaceIds', req)
-      self._update_session(response)
-      if response.reply.get('Error'):
-        raise gorpc.AppError(
-            response.reply['Error'], 'VTGate.ExecuteBatchKeyspaceIds')
-      for reply in response.reply['List']:
-        fields = []
-        conversions = []
-        results = []
-        rowcount = 0
-
-        for field in reply['Fields']:
-          fields.append((field['Name'], field['Type']))
-          conversions.append(field_types.conversions.get(field['Type']))
-
-        for row in reply['Rows']:
-          results.append(tuple(_make_row(row, conversions)))
-
-        rowcount = reply['RowsAffected']
-        lastrowid = reply['InsertId']
-        rowsets.append((results, rowcount, lastrowid, fields))
-    except gorpc.GoRpcError as e:
-      self.logger_object.log_private_data(bind_variables_list)
-      raise self._convert_exception(
-          e, sql_list, keyspace_ids_list,
-          keyspace='', tablet_type=tablet_type)
-    except Exception:
-      logging.exception('gorpc low-level error')
-      raise
+    keyspace_ids_rowsets = make_execute_batch_call(
+        keyspace_ids_query_list, is_keyspace_ids=True)
+    shards_rowsets = make_execute_batch_call(
+        shards_query_list, is_keyspace_ids=False)
+    keyspace_ids_iter = iter(keyspace_ids_rowsets)
+    shards_iter = iter(shards_rowsets)
+    for is_keyspace_ids in is_keyspace_ids_list:
+      if is_keyspace_ids:
+        rowsets.append(keyspace_ids_iter.next())
+      else:
+        rowsets.append(shards_iter.next())
```

Comment:
We will be implementing unsharded tables using shards='0' and sharded
tables usually using keyspace_ids. I can see this as being a non-edge case
in flows like this:

with transaction_manager.ExecuteBatchContext as execute_batch_context:
    for data in data_list:
      enqueue_an_unsharded_table_query(execute_batch_context,
data.unsharded_part)
      enqueue_a_sharded_table_query(execute_batch_context,
data.sharded_part)

I expect that developers could break it up into two passes, too. Either
support it or raise an exception if a batch contains some of each.

I have a clean-up commit coming after I test it.



Dean Yasuda

On Tue, Sep 22, 2015 at 10:51 AM, Alain Jobart <notifications@github.com>
wrote:

> In py/vtdb/vtgatev2.py
> <https://github.com/youtube/vitess/pull/1120#discussion_r40119187>:
>
> > -          e, sql_list, keyspace_ids_list,
> > -          keyspace='', tablet_type=tablet_type)
> > -    except Exception:
> > -      logging.exception('gorpc low-level error')
> > -      raise
> > +    keyspace_ids_rowsets = make_execute_batch_call(
> > +        keyspace_ids_query_list, is_keyspace_ids=True)
> > +    shards_rowsets = make_execute_batch_call(
> > +        shards_query_list, is_keyspace_ids=False)
> > +    keyspace_ids_iter = iter(keyspace_ids_rowsets)
> > +    shards_iter = iter(shards_rowsets)
> > +    for is_keyspace_ids in is_keyspace_ids_list:
> > +      if is_keyspace_ids:
> > +        rowsets.append(keyspace_ids_iter.next())
> > +      else:
> > +        rowsets.append(shards_iter.next())
>
> Is that a use case we actually have? I mean, it's more code, that we have
> to test and maintain, not sure about the gain?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/youtube/vitess/pull/1120/files#r40119187>.
>